Pass over the accessibility tree section (#629)
browserengineering/book: book/accessibility.md

Screen readers
==============

Zoom, dark mode, and focus indicators help users with difficulty
seeing fine details, but if the user can't see the screen at all,^[The
original motivation of screen readers was for blind users, but it's
also sometimes useful for situations where the user shouldn't be
looking at the screen (such as driving), or for devices with no
screen.] screen-reader software is typically used instead. The name
kind of explains it all: screen-reader software reads the text on the
screen, so that users know what it says without having to see it.

So: what should we say to the user? There are basically two big
challenges we must overcome.

> First, web pages contain visual hints besides text, that we need to

Reviewer: ```suggestion
First, web pages contain visual hints besides text that we need to
```

reproduce for screen-reader users. For example, when focus is on an
`<input>` or `<button>` element, the screen-reader needs to say so,
since these users won't see the light blue or orange background.

And second, when listening to a screen-reader, the user must be able
to direct the browser to the part of the page that interests
them.[^fast] For example, the user might want to skip headers and
navigation menus, or even skip most of the page until they get to a
paragraph of interest. But once they've reached the part of the page
of interest to them, they may want it read to them, and if some
sentence or phrase is particularly complex, they may want the
screen-reader to re-read it.

> You see an example[^imagine] of screen-reader navigation in this talk,

Reviewer: ```suggestion
You can see an example[^imagine] of screen-reader navigation in this talk,
```

specifically the segment from 2:36--3:54:[^whole-talk]

<iframe width="560" height="315" src="https://www.youtube-nocookie.com/embed/qi0tY60Hd6M?start=159" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

[^whole-talk]: The whole talk is recommended; it has great examples of
    using accessibility technology.

> [^fast]: Though though many people who rely on screen-readers learn to

Reviewer: ```suggestion
[^fast]: Though many people who rely on screen-readers learn to
```

    listen to *much* faster speech, it's still a less informationally
    dense medium than vision.

[^imagine] I encourage you to test out your operating system's
    built-in screen reader to get a feel for what screen-reader
    navigation is like. On macOS, type Cmd-Fn-F5 to turn on Voice
    Over; on Windows, type Win-Ctrl-Enter or Win-Enter to start
    Narrator. Both are largely used via keyboard shortcuts that you
    can look up.
    

> To support all this, screen readers structure the page as a tree. The

Reviewer: ```suggestion
To support all this, browsers structure the page as a tree and use it to respond to screen reader APIs (these APIs are generally usually implemented outside the browser). The
```

higher levels of the tree represent items like paragraphs, headings,
or navigation menus, while lower levels represent text, links, or
buttons.

This probably sounds a lot like HTML---and it is quite similar! But,
just like the HTML tree does not exactly match the layout tree,
there's not an exact match with this tree either. For example, some
HTML elements (like `<div>`) group content for styling that is
meaningless to screen-reader users. Alternatively, some HTML elements
may be invisible on the screen,[^invisible-example] but relevant to
screen-reader users. The browser therefore builds a separate
[accessibility tree][at] to support screen-reader navigation.

[at]: https://developer.mozilla.org/en-US/docs/Glossary/Accessibility_tree

[^invisible-example]: For example, using `opacity:0`. There are
several other ways in real browsers that elements can be made
invisible, such as with the `visibility` or `display` CSS properties.

> The accessibility tree is built in a rendering phase just after layout:

Reviewer: ```suggestion
Let's implement the accessibility tree. It is built in a rendering phase just after layout:
```

node based on its tag name, or from the special `role` attribute if
that exists:

[^standard] Roles and default roles are are specified in the
[WAI ARIA standard][aria-roles].

[aria-roles]: https://www.w3.org/TR/wai-aria-1.2/#introroles

``` {.python}
class AccessibilityNode:
    def __init__(self, node):
        # ...
        if isinstance(node, Text):
            if node.parent.tag == "a":
                self.role = "link"
            elif is_focusable(node.parent):
                self.role = "focusable text"
            else:
                self.role = "StaticText"
        else:
            if "role" in node.attributes:
                self.role = node.attributes["role"]
            elif node.tag == "a":
                self.role = "link"
            elif node.tag == "input":
                self.role = "textbox"
            elif node.tag == "button":
                self.role = "button"
            elif node.tag == "html":
                self.role = "document"
            elif is_focusable(node):
                self.role = "focusable"
            else:
                self.role = "none"
```

To build the accessibility tree, we just recursively walk the HTML

> tree, adding all nodes whose role isn't `none`:

Reviewer: ```suggestion
tree, adding all nodes whose role isn't `none` (note that a regular `<div>` is therefore skipped, for example):
```

``` {.python}
class AccessibilityNode:
    def build(self):
        for child_node in self.node.children:
            self.build_internal(child_node)

    def build_internal(self, node):
        child = AccessibilityNode(node)
        if child.role != "none":
            self.children.append(child)
            parent = child
        else:
            parent = self
        for child_node in node.children:
            parent.build_internal(child_node)
```

> The user can now direct the screen-reader to walks up or down this

Reviewer: ```suggestion
The user can now direct the screen-reader to walk up or down this
```

of interest to them, they may want it read to them, and if some
sentence or phrase is particularly complex, they may want the
screen-reader to re-read it.

You can see an example[^imagine] of screen-reader navigation in this
talk, specifically the segment from 2:36--3:54:[^whole-talk]

<iframe width="560" height="315" src="https://www.youtube-nocookie.com/embed/qi0tY60Hd6M?start=159" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

[^whole-talk]: The whole talk is recommended; it has great examples of
    using accessibility technology.

[^fast]: Though many people who rely on screen-readers learn to listen
    to *much* faster speech, it's still a less informationally dense
    medium than vision.

[^imagine] I encourage you to test out your operating system's
    built-in screen reader to get a feel for what screen-reader
    navigation is like. On macOS, type Cmd-Fn-F5 to turn on Voice
    Over; on Windows, type Win-Ctrl-Enter or Win-Enter to start
    Narrator. Both are largely used via keyboard shortcuts that you
    can look up.
    

> To support all this, browers structure the page as a tree and use that

Reviewer: ```suggestion
To support all this, browsers structure the page as a tree and use that
```

    os.remove(SPEECH_FILE)
```

::: {.quirk}

You may need to adjust the `SPEECH_FILE` path to fit your system
better. If you have trouble importing any of the libraries, you may
need to consult the [`gtts`][gtts] or [`playsound`][playsound]
documentation. If you can't get these libraries working, you can skip
the "play the sound out loud" aspect of this chapter by replacing
`speak_text` with something like:

``` {.python.example}
def speak_text(text):
    print("SPEAK:", text)
```

Then you can see what the browser *would have* spoken out loud in the
standard output.

:::

> To start with, we'll want a keybinding that turns the screen-reader on

Reviewer: ```suggestion
To start with, we'll want a key binding that turns the screen-reader on
```

and off. While real operating systems typically use more obscure
shortcuts, I'll use `Ctrl-A` to turn on the screen-reader:

``` {.python}
if __name__ == "__main__":
    while True:
        if sdl2.SDL_PollEvent(ctypes.byref(event)) != 0:
            # ...
                if ctrl_down:
                    # ...
                    elif event.key.keysym.sym == sdl2.SDLK_a:
                        browser.toggle_accessibility()            
```

The `toggle_accessibility` method tells the `Tab` that accessibility
is on:

``` {.python}
class Browser:
    def __init__(self):
        self.accessibility_is_on = False

    def toggle_accessibility(self):
        self.lock.acquire(blocking=True)
        self.accessibility_is_on = not self.accessibility_is_on
        active_tab = self.tabs[self.active_tab]
        task = Task(active_tab.toggle_accessibility)
        active_tab.task_runner.schedule_task(task)
        self.lock.release()
```

The `Tab`, in turn, executes the `speak_task` method if accessibility

> is on, which is actually produce sound:

Reviewer: ```suggestion
is on, which is what actually produces sound:
```
